File: cpp/myLeetCode/Greedy/Interval_problem/435. Non-overlapping Intervals.cpp
// 435. Non-overlapping Intervals
// https://leetcode.com/problems/non-overlapping-intervals/

#include <iostream>
#include <vector>
#include <algorithm>

struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};

bool compareIntervals(const Interval& a, const Interval& b) {
    return a.start < b.start;
}

int eraseOverlapIntervals(std::vector<Interval>& intervals) {
    if (intervals.empty()) {
        return 0;
    }

    std::sort(intervals.begin(), intervals.end(), compareIntervals);

    int count = 0;
    int prevEnd = intervals[0].end;

    for (int i = 1; i < intervals.size(); i++) {
        if (intervals[i].start < prevEnd) {
            count++;
            prevEnd = std::min(prevEnd, intervals[i].end);
        } else {
            prevEnd = intervals[i].end;
        }
    }

    return count;
}

int main() {
    std::vector<Interval> intervals{{1,2}, {2,3}, {3,4}, {1,3}};
    std::cout << "Minimum number of intervals to remove: " << eraseOverlapIntervals(intervals) << std::endl;

    std::vector<Interval> intervals2{{1,2}, {1,2}, {1,2}};
    std::cout << "Minimum number of intervals to remove: " << eraseOverlapIntervals(intervals2) << std::endl;

    std::vector<Interval> intervals3{{1,2}, {2,3}};
    std::cout << "Minimum number of intervals to remove: " << eraseOverlapIntervals(intervals3) << std::endl;

    return 0;
}

File: C++/代码随想录/Greedy/Interval_problem/435. Non-overlapping Intervals.py
def eraseOverlapIntervals(intervals):
    if not intervals:
        return 0

    # 根据每个区间的结束时间对区间进行排序
    intervals.sort(key=lambda x: x[1])

    # 初始化
    end = intervals[0][1]
    count = 0

    for i in range(1, len(intervals)):
        # 如果当前区间的开始时间小于上一个区间的结束时间
        if intervals[i][0] < end:
            # 需要移除一个区间
            count += 1
        else:
            # 更新结束时间
            end = intervals[i][1]

    return count

# 测试用例
print(eraseOverlapIntervals([[1,2],[2,3],[3,4],[1,3]]))  # 应该输出 1
print(eraseOverlapIntervals([[1,2],[1,2],[1,2]]))        # 应该输出 2
print(eraseOverlapIntervals([[1,2],[2,3]]))              # 应该输出 0
